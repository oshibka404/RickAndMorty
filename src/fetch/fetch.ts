function getRequestUrl(query: string, page: number) : string {
    return query 
        ? `https://rickandmortyapi.com/api/character/?name=${query}&page=${page}`
        : `https://rickandmortyapi.com/api/character/?page=${page}`;
}

const responseCache : Record<string, CharacterResponseData[]> = {};

export async function fetchCharacters(query: string, page: number) {
    if (responseCache[query] && responseCache[query][page]) {
        return Promise.resolve(responseCache[query][page]);
    }
    const response = await fetch(getRequestUrl(query, page));
    if (!response.ok) {
        return Promise.reject(
            response.status === 404
                ? await response.json()
                : {error: response.statusText}
        );
    }
    const data = await response.json();
    responseCache[query] = responseCache[query] || [];
    return responseCache[query][page] = data;
}

export interface CharacterResponseData {
    info: ResponseInfo;
    results: Character[];
}

export interface ResponseInfo {
    count: number;
    pages: number;
    next: string;
    prev: string;
}

export interface Character {
    id: number;
    name: string;
    status: string;
    image: string;
}
