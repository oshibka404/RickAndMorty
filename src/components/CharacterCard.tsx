import React from 'react';
import {StyleSheet, View, Image, Text} from 'react-native';
import {Character} from '../fetch/fetch';

interface Props {
    item: Character;
}

export function CharacterCard({item}: Props) {
    return (
        <View style={styles.character}>
            <Image
                source={{uri: item.image}}
                style={styles.characterImage}
            />
            <View style={styles.characterInfo}>
                <Text style={styles.characterName}>{item.name}</Text>
                <Text>{item.status}</Text>
            </View>
        </View>
    );
}

const styles = StyleSheet.create({
    character: {
        flexDirection: 'row',
        paddingBottom: 10,
        paddingHorizontal: 5
    },
    characterName: {
        fontSize: 16
    },
    characterImage: {
        width: 100,
        height: 100,
    },
    characterInfo: {
        flexDirection: 'column',
        paddingHorizontal: 5
    }
});
