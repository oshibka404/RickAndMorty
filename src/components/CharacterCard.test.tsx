import React from 'react';
import renderer from 'react-test-renderer';
import {CharacterCard} from './CharacterCard';
import { CharacterResponseData } from '../fetch/fetch';
const {data}: {data: CharacterResponseData} = require('../__mocks__/response-mock');

describe('Character card', () => {
  test('Renders correctly', () => {
    const tree = renderer.create(
      <CharacterCard
        item={data.results[0]}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
