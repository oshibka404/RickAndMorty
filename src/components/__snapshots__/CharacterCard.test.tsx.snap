// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character card Renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "paddingBottom": 10,
      "paddingHorizontal": 5,
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
      }
    }
    style={
      Object {
        "height": 100,
        "width": 100,
      }
    }
  />
  <View
    style={
      Object {
        "flexDirection": "column",
        "paddingHorizontal": 5,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 16,
        }
      }
    >
      Rick Sanchez
    </Text>
    <Text>
      Alive
    </Text>
  </View>
</View>
`;
